src/aspire/denoise/covar2d_ctf.py

```diff
@@ -16,31 +28,142 @@ def get_mean_ctf(self, coeffs, ctf_fb, ctf_idx):
-        :return: The covariance matrix of coefficients for all images.
+        b = np.zeros((self.basis.basis_count, 1), dtype=self.as_type)
 
-        """
-        pass
+        A = blk_diag_zeros(blk_diag_partition(ctf_fb[0]), dtype=self.as_type)
+        for k in np.unique(ctf_idx[:]).T:
+            coeff_k = coeffs[:, ctf_idx == k]
+            weight = np.size(coeff_k, 1)/np.size(coeffs, 1)
+            mean_coeff_k = self.get_mean(coeff_k)
+            ctf_fb_k = ctf_fb[k]
+            ctf_fb_k_t = blk_diag_transpose(ctf_fb_k)
+            b = b + weight*blk_diag_apply(ctf_fb_k_t, mean_coeff_k)
+            A = blk_diag_add(A, blk_diag_mult(weight, blk_diag_mult(ctf_fb_k_t, ctf_fb_k)))
 
-    def _shrink(self, covar_b_coeff, noise_variance, method=None):
-        """
-        Shrink covariance matrix
-        :param covar_b_coeff: Outer products of the mean-subtracted images
-        :param noise_variance: Noise variance
-        :param method: One of None/'frobenius_norm'/'operator_norm'/'soft_threshold'
-        :return: Shrunk covariance matrix
-        """
-        pass
+        mean_coeff = blk_diag_solve(A, b)
+
+        return mean_coeff
+
+    def get_covar_ctf(self, coeffs, ctf_fb, ctf_idx, mean_coeff=None, noise_var=1, covar_est_opt=None):
+
+        def identity(x):
+            return x
+
+        if covar_est_opt is None:
+            covar_est_opt = {'shrinker': 'none', 'verbose': 0, 'max_iter': 250, 'iter_callback': [],
+                             'store_iterates': False, 'rel_tolerance': 1e-12, 'precision': 'float64',
+                             'preconditioner': 'identity'}
+
+        if mean_coeff is None:
+            mean_coeff = self.get_mean_ctf(coeffs, ctf_fb, ctf_idx)
+
+        block_partition = blk_diag_partition(ctf_fb[0])
+        b_coeff = blk_diag_zeros(block_partition, dtype=self.as_type)
+        b_noise = blk_diag_zeros(block_partition, dtype=self.as_type)
+        A = []
+        for k in range(0, len(ctf_fb)):
+            A.append(blk_diag_zeros(block_partition, dtype=self.as_type))
+
+        M = blk_diag_zeros(block_partition, dtype=self.as_type)
+
+        for k in np.unique(ctf_idx[:]):
+
+            coeff_k = coeffs[:, ctf_idx == k]
+            weight = np.size(coeff_k, 1)/np.size(coeffs, 1)
+
+            ctf_fb_k = ctf_fb[k]
+            ctf_fb_k_t = blk_diag_transpose(ctf_fb_k)
+            mean_coeff_k = blk_diag_apply(ctf_fb_k, mean_coeff)
+            covar_coeff_k = self.get_covar(coeff_k, mean_coeff_k)
+
+            b_coeff = blk_diag_add(b_coeff, blk_diag_mult(ctf_fb_k_t,
+                blk_diag_mult(covar_coeff_k, blk_diag_mult(ctf_fb_k, weight))))
+
+            b_noise = blk_diag_add(b_noise, blk_diag_mult(weight,
+                blk_diag_mult(ctf_fb_k_t, ctf_fb_k)))
+
+            A[k] = blk_diag_mult(ctf_fb_k_t, blk_diag_mult(ctf_fb_k, np.sqrt(weight)))
+
+            M = blk_diag_add(M, A[k])
+
+        if covar_est_opt['shrinker'] == 'none':
```

Comment:
Don't have 'none' as a value. Use the Python None.

Reply:
fixed in new commit.